### PR TITLE
Add eth light client

### PR DESCRIPTION
**Pull request by 2841156671**
* [x] init eth light client
* [x] client types 
* [x] errors
* [x] todo

### CODE REVIEW

File: modules/tibc/light-clients/09-eth/types/client_state.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	fmt "fmt"
+	host "github.com/bianjieai/tibc-go/modules/tibc/core/24-host"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/client_state.go:11]: make format

File: modules/tibc/light-clients/09-eth/types/header.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
+package types
+
+import (
+	fmt "fmt"
+	"github.com/ethereum/go-ethereum/core/types"
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/header.go:5]: ditto.

File: modules/tibc/light-clients/09-eth/types/snapshot.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+type snapshot struct {
+	cdc   codec.BinaryMarshaler
+	store sdk.KVStore
+	// Block number where the snapshot was created
+	Number uint64
+	// Set of authorized signers at this moment
+	Signers map[common.Address]struct{} `json:"signers"`
+	// Set of authorized validators at this moment
+	Validators map[common.Address]struct{}
+	Recents    map[uint64]common.Address
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/snapshot.go:27]: Are these required?

File: modules/tibc/light-clients/09-eth/types/client_state.go (hunk 1 above)
Comment (2841156671) [reply modules/tibc/light-clients/09-eth/types/client_state.go:11]: ok

File: modules/tibc/light-clients/09-eth/types/header.go (hunk 2 above)
Comment (2841156671) [reply modules/tibc/light-clients/09-eth/types/header.go:5]: ok

File: modules/tibc/light-clients/09-eth/types/snapshot.go (hunk 3 above)
Comment (2841156671) [reply modules/tibc/light-clients/09-eth/types/snapshot.go:27]: snapshot is not clear if it's needed，next step i will make sure.

File: modules/tibc/light-clients/09-eth/types/update.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	consState *ConsensusState,
+	header Header,
+
+) error {
+	if err := header.ValidateBasic(); err != nil {
+		return err
+	}
+	//if clientState.Header.Hash() != header.ToEthHeader().ParentHash {
+	//	return errors.New("parent hash not same")
+	//}
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/update.go:80]: delete

File: modules/tibc/light-clients/09-eth/types/update.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If  verify succeeds, save consensusState first . this store is header_index
+	consensusStatetmp, err := cdc.MarshalInterface(consensusState)
+	if err != nil {
+		return nil, nil, err
+	}
+	store.Set(host.ConsensusStateIndexKey(consensusState.Header.Hash()), consensusStatetmp)
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/update.go:52]: Not add this key in `host`?

File: modules/tibc/light-clients/09-eth/types/header.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+func (h *EthHeader) Hash() (hash common.Hash) {
+	return rlpHash(h)
+}
+
+func (h Header) ValidateBasic() error {
+	// Ensure that the header's extra-data section is of a reasonable size
+	if uint64(len(h.Extra)) > params.MaximumExtraDataSize {
+		return fmt.Errorf("extra-data too long: %d > %d", len(h.Extra), params.MaximumExtraDataSize)
+	}
+	if h.Time > uint64(time.Now().Unix()+allowedFutureBlockTimeSeconds) {
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/header.go:53]: never use time.Now(), use `context.blockTime()`.

File: modules/tibc/light-clients/09-eth/types/header.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+	ethHeader := parent.Header.ToEthHeader()
+	if !bytes.Equal(ethHeader.Hash().Bytes(), header.ToEthHeader().ParentHash.Bytes()) {
+		return fmt.Errorf("SyncBlockHeader, parent header is not right. Header: %s", header.String())
+	}
+	//verify whether extra size validity
+	if uint64(len(header.Extra)) > params.MaximumExtraDataSize {
+		return sdkerrors.Wrap(ErrExtraLenth, fmt.Errorf("SyncBlockHeader, SyncBlockHeader extra-data too long: %d > %d, header: %s", len(header.Extra), params.MaximumExtraDataSize, header.String()).Error())
+	}
+	// Verify the header's timestamp
+	if header.Time > uint64(time.Now().Unix()+allowedFutureBlockTimeSeconds) {
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/header.go:170]: ditto.

File: modules/tibc/light-clients/09-eth/types/client_state.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+func (m ClientState) Status(
+	ctx sdk.Context,
+	store sdk.KVStore,
+	cdc codec.BinaryMarshaler,
+) exported.Status {
+	onsState, err := GetConsensusState(store, cdc, m.GetLatestHeight())
+	if err != nil {
+		return exported.Unknown
+	}
+	if onsState.Timestamp+uint64(allowedFutureBlockTimeSeconds) < uint64(ctx.BlockTime().Nanosecond()) {
```
Comment (chengwenxi) [inline modules/tibc/light-clients/09-eth/types/client_state.go:76]: allowedFutureBlockTimeSeconds -> delayTime

File: modules/tibc/core/exported/client.go [hunk 9]
```diff
@@ -20,6 +20,8 @@ const (
 	// BSC is the client type for a bianance smart chain client.
 	BSC string = "008-bsc"
 
+	ETH string = "009-eth"
+
 	// Fabric is the client type for a hyperledge fabric client.
 	Fabric string = "009-fabric"
```
Comment (aofengli) [inline modules/tibc/core/exported/client.go:26]: ETH string = "009-eth" 
Fabric string = "009-fabric"
Is it conflicting?

File: modules/tibc/light-clients/09-eth/types/client_state.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+func (m ClientState) Validate() error {
+	return m.Header.ValidateBasic()
+}
+
+func (m ClientState) GetDelayTime() uint64 {
+	return 0
+}
+
+func (m ClientState) GetDelayBlock() uint64 {
+	return 0
```
Comment (aofengli) [inline modules/tibc/light-clients/09-eth/types/client_state.go:40]: Should it be a non-zero value here?
@chengwenxi

File: modules/tibc/core/exported/client.go (hunk 9 above)
Comment (2841156671) [reply modules/tibc/core/exported/client.go:26]: ETH string = "010-eth"

File: modules/tibc/light-clients/09-eth/types/client_state.go (hunk 10 above)
Comment (2841156671) [reply modules/tibc/light-clients/09-eth/types/client_state.go]: clientState.BloclDelay

**Review by aofengli — APPROVED** [review]

**Review by chengwenxi — APPROVED** [review]

**Pull request merged**